File: app/src/main/java/com/stevenfrew/beatprompter/bluetooth/AcceptConnectionsFromClientsThread.java
package com.stevenfrew.beatprompter.bluetooth;

import android.bluetooth.BluetoothAdapter;
import android.bluetooth.BluetoothServerSocket;
import android.bluetooth.BluetoothSocket;
import android.util.Log;

import java.io.IOException;

public class AcceptConnectionsFromClientsThread extends Thread
{
    private BluetoothServerSocket mmServerSocket=null;
    private boolean mStop=false;
    private final Object mSocketNullLock=new Object();
    private BluetoothAdapter mBluetoothAdapter;

    AcceptConnectionsFromClientsThread(BluetoothAdapter bluetoothAdapter)
    {
        mBluetoothAdapter=bluetoothAdapter;
    }

    public void run() {
        // Keep listening until exception occurs or a socket is returned
        while (!mStop)
        {
            try {
                BluetoothServerSocket serverSocket;
                BluetoothSocket acceptedSocket;
                synchronized(mSocketNullLock)
                {
                    if(mmServerSocket==null)
                    {
                        try {
                            // MY_UUID is the app's UUID string, also used by the server code
                            mmServerSocket = mBluetoothAdapter.listenUsingRfcommWithServiceRecord(BluetoothConstants.APP_BLUETOOTH_NAME, BluetoothConstants.APP_BLUETOOTH_UUID);
                        } catch (IOException e) {
                            Log.e(BluetoothManager.BLUETOOTH_TAG, "Error creating Bluetooth socket.",e);
                        }
                    }
                    serverSocket=mmServerSocket;
                }

                if(serverSocket!=null) {
                    acceptedSocket = serverSocket.accept(2000);
                    // If a connection was accepted
                    if (acceptedSocket != null) {
                        // Do work to manage the connection (in a separate thread)
                        BluetoothManager.handleConnectionFromClient(acceptedSocket);
                    }
                }
            } catch (IOException e) {
                //Log.e(BLUETOOTH_TAG, "Failed to accept new Bluetooth connection.",e);
            }
        }
    }

    /** Will cancel the listening socket, and cause the thread to finish */
    public void stopListening()
    {
        mStop=true;
        try {
            mmServerSocket.close();
        } catch (IOException e) {
            Log.e(BluetoothManager.BLUETOOTH_TAG, "Failed to close Bluetooth listener socket.",e);
        }
        finally
        {
            synchronized (mSocketNullLock)
            {
                mmServerSocket=null;
            }
        }
    }
}

File: app/src/main/kotlin/com/stevenfrew/beatprompter/bluetooth/AcceptConnectionsFromClientsThread.kt
package com.stevenfrew.beatprompter.bluetooth

import android.bluetooth.BluetoothAdapter
import android.bluetooth.BluetoothServerSocket
import android.util.Log
import com.stevenfrew.beatprompter.BeatPrompterApplication
import java.io.IOException

class AcceptConnectionsFromClientsThread internal constructor(private val mBluetoothAdapter: BluetoothAdapter) : Thread() {
    private var mmServerSocket: BluetoothServerSocket? = null
    private var mStop = false
    private val mSocketNullLock = Any()

    override fun run() {
        // Keep listening until exception occurs or a socket is returned
        while (!mStop) {
            try {
                var serverSocket: BluetoothServerSocket?=null
                synchronized(mSocketNullLock) {
                    if (mmServerSocket == null) {
                        try {
                            // MY_UUID is the app's UUID string, also used by the server code
                            mmServerSocket = mBluetoothAdapter.listenUsingRfcommWithServiceRecord(BeatPrompterApplication.APP_NAME, BluetoothManager.BLUETOOTH_UUID)
                        } catch (e: IOException) {
                            Log.e(BluetoothManager.BLUETOOTH_TAG, "Error creating Bluetooth socket.", e)
                        }
                    }
                    serverSocket = mmServerSocket
                }

                // If a connection was accepted
                // Do work to manage the connection (in a separate thread)
                serverSocket?.accept(2000)?.also{
                    BluetoothManager.handleConnectionFromClient(it)
                }
            } catch (e: IOException) {
                //Log.e(BLUETOOTH_TAG, "Failed to accept new Bluetooth connection.",e);
            }

        }
    }

    /** Will cancel the listening socket, and cause the thread to finish  */
    fun stopListening() {
        mStop = true
        synchronized(mSocketNullLock) {
            try {
                mmServerSocket?.close()
            } catch (e: IOException) {
                Log.e(BluetoothManager.BLUETOOTH_TAG, "Failed to close Bluetooth listener socket.", e)
            } finally {
                mmServerSocket = null
            }
        }
    }
}
